When using PARAM_DEFINE_* macros in C++, get "designator order does not match declaration order"

Compiler: g++ 11.3.0

Expected behavior: `PARAM_DEFINE_*` macros generate compiling C++ code.

Actual behavior: The compile fails with the error:

```
param.h:199:9: error: designator order for field ‘param_s::id’ does not match declaration order in ‘param_t’ {aka ‘param_s’}
```

This appears to be because in C++20, the order of designated initializers has to match the declaration order.  See:

> Note: out-of-order designated initialization, nested designated initialization, mixing of designated initializers and regular initializers, and designated initialization of arrays are all supported in the [C programming language](https://en.cppreference.com/w/c/language/struct_initialization), but are not allowed in C++.

From https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers


It seems like this can be fixed just by re-ordering the initializers and inserting NULL initializers as needed in the `PARAM_DEFINE_*` macros.

---

Reproducible example: In a new project directory:

Clone libcsp into lib/csp and libparam into lib/param (as done here: https://github.com/spaceinventor/csh)

meson.build file:
```
project('TestProgram',
	'cpp',
	subproject_dir: 'lib'
)

csp_dep = dependency('csp', fallback: ['csp', 'csp_dep'], required: true)
param_dep = dependency('param', fallback: ['param', 'param_dep'], required: true)

main = executable(
	'main',
	[ 'src/main.cpp' ],
	dependencies: [csp_dep, param_dep]
)
```

src/main.cpp file:
```
#include <param/param.h>

static uint8_t _ch_on[12];
PARAM_DEFINE_REMOTE_DYNAMIC(20, ch_on, 3, PARAM_TYPE_UINT8, 12, 0, PM_CONF, &_ch_on, "");

int main() {
  return 0;
}
```

--- a/include/param/param.h
+++ b/include/param/param.h
@@ -124,19 +124,19 @@ typedef struct param_s {
 	__attribute__((aligned(1))) \
 	__attribute__((used)) \
 	param_t _name = { \
-		.vmem = NULL, \
-		.node = 0, \
 		.id = _id, \
+		.node = 0, \
 		.type = _type, \
+		.mask = _flags, \
 		.name = #_name, \
+		.unit = _unit, \
+		.docstr = _docstr, \
+		.addr = _physaddr, \
+		.vmem = NULL, \
 		.array_size = _array_count < 1 ? 1 : _array_count, \
 		.array_step = _array_step, \
-		.mask = _flags, \
-		.unit = _unit, \
 		.callback = _callback, \
 		.timestamp = &_timestamp_##_name, \
-		.addr = _physaddr, \
-		.docstr = _docstr, \
 	}
 
 #define PARAM_DEFINE_STATIC_VMEM(_id, _name, _type, _array_count, _array_step, _flags, _callback, _unit, _vmem_name, _vmem_addr, _docstr) \
@@ -146,19 +146,19 @@ typedef struct param_s {
 	__attribute__((aligned(1))) \
 	__attribute__((used)) \
 	param_t _name = { \
-		.node = 0, \
 		.id = _id, \
+		.node = 0, \
 		.type = _type, \
-		.name = #_name, \
-		.array_size = _array_count < 1 ? 1 : _array_count, \
-		.array_step = _array_step, \
 		.mask = _flags, \
-		.callback = _callback, \
-		.timestamp = &_timestamp_##_name, \
+		.name = #_name, \
 		.unit = _unit, \
+		.docstr = _docstr, \
 		.addr = (void *) _vmem_addr, \
 		.vmem = &vmem_##_vmem_name, \
-		.docstr = _docstr, \
+		.array_size = _array_count < 1 ? 1 : _array_count, \
+		.array_step = _array_step, \
+		.callback = _callback, \
+		.timestamp = &_timestamp_##_name, \
 	}
 
 #define PARAM_REMOTE_NODE_IGNORE 16382
@@ -170,32 +170,38 @@ typedef struct param_s {
 	__attribute__((aligned(1))) \
 	__attribute__((used)) \
 	param_t _name = { \
-		.node = _node, \
 		.id = _id, \
+		.node = _node, \
 		.type = _type, \
+		.mask = _flags, \
+		.name = (char *) #_name, \
+		.unit = NULL, \
+		.docstr = _docstr, \
+		.addr = _physaddr, \
+		.vmem = NULL, \
 		.array_size = _array_count < 1 ? 1 : _array_count, \
 		.array_step = _array_step, \
-		.name = (char *) #_name, \
-		.mask = _flags, \
-		.addr = _physaddr, \
+		.callback = NULL, \
 		.timestamp = &_timestamp_##_name, \
-		.docstr = _docstr, \
 	};
 
 #define PARAM_DEFINE_REMOTE_DYNAMIC(_id, _name, _node, _type, _array_count, _array_step, _flags, _physaddr, _docstr) \
 	; /* Catch const param defines */ \
 	uint32_t _timestamp_##_name = 0; \
 	param_t _name = { \
-		.node = _node, \
 		.id = _id, \
+		.node = _node, \
 		.type = _type, \
+		.mask = _flags, \
+		.name = (char *) #_name, \
+		.unit = NULL, \
+		.docstr = _docstr, \
+		.addr = _physaddr, \
+		.vmem = NULL, \
 		.array_size = _array_count < 1 ? 1 : _array_count, \
 		.array_step = _array_step, \
-		.name = (char *) #_name, \
-		.mask = _flags, \
-		.addr = _physaddr, \
+		.callback = NULL, \
 		.timestamp = &_timestamp_##_name, \
-		.docstr = _docstr, \
 	};
 
 /* Native getter functions, will return native types */

